docs/designs/allreduce.md

```diff
@@ -1,14 +1,215 @@
+fault-tolerant allreduce. The initial implementation will contain an experimental Python binding for NCCL that is
+fault-tolerant and Kubernetes-native. This will include but not limited to the following objectives (more details to be disclosed later once
+the implementation has been open-sourced):
+
+* Fault-tolerant: if any of the worker pod fails, the [NCCL communicator](https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/usage/communicators.html)
+can be reconstructed. The allreduce operation continues as long as there's at least one healthy worker pod.
+* Elastic: the number of worker pods can be dynamically added if there are enough computational resources available.
+The NCCL communicator will be reconstructed and the ranks can be re-assigned as the number of worker pods changes.
+
+The interface for allreduce would look like the following:
+
+```python
+num_epochs = 2
+num_batches = 10
+data_loader = DataLoader(num_batches)
+communicator = AllreduceCommunicator()
+
+for _ in range(num_epochs):
+    for features, labels in data_loader:
+        outputs = model(features)
+        loss = loss_fn(outputs, labels)
+        grads = calculate_gradients(loss, model)
+        res, averaged_grads = communicator.average_gradients(grads)
+        if res == SUCCESS:
+            update_model(averaged_grads)
+        elif res == FAILED:
+            continue
+```
+
+We also provide API to broadcast all model parameters from one of the worker pods to all other active worker pods.
+This would be useful to initialize the model parameters for a restarted or new joining worker pod. The interface would
+look like the following:
+
+```python
+communicator = BroadcastCommunicator()
+init_local_model(worker_0_ip)
+communicator.broadcast(worker_ips, get_model_params(worker_0_ip)
+```
+
+### Allreduce-based Training in ElasticDL
+
+#### Gradients Averaging and Model Updating
+
+For training based on parameter servers, gradients calculation and model updating include the following steps:
+
+1. Send the gradients calculated locally from each worker to parameter servers.
+1. Calculate the average of all the received gradients on master.
+1. Update the model in master.
+
+On contrary, in allreduce-based training, each worker in ElasticDL calculates gradients locally and then calculates
+the average of gradients across all workers using collective communication via `AllreduceCommunicator.average_gradients()`
+that we mentioned in the previous section. The main differences are the following:
+
+1. Gradients from each worker are not sent to master.
+1. The average of gradients across all workers is calculated locally on each worker.
+1. The model is updated directly on each worker and each worker has the exact same copy of the model.
+
+Below is the pseudo-code for this process on each worker:
+
+```python
+communicator = AllreduceCommunicator()
```

Comment:
Should we pass the ips of other worker pods into the AllReduceCommunicator?

Reply:
Good question. I don't think we need to pass IPs for allreduce since the shared storage in the implementation would keep track of the list of IPs. I just added a note on this.

cc @zw0610 to double check.

Reply:
I checked MPI and NCCL: the _communicator_ in the sense of these communication libraries ask no IPs during the creation procedure nor the allreduce procedure.

The proto-implementation uses shared storage, keeping track of the list of IPs, which represent workers. So the `AllreduceCommunicator` in this context would need the IP of itself (instead of others) to register on the shared storage and build a consensus of all the workers.

Let me know if there is any other thoughts.

Reply:
@zw0610 Thanks! The IP of the `AllreduceCommunicator` itself should be automatically detected by `AllreduceCommunicator` though so users of the lib would not need to worry about this.